cpp/src/dictionary/dictionary_factories.cu

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/dictionary/dictionary_factories.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/gather.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+
+namespace cudf
+{
+namespace detail
+{
+namespace
+{
+
+// Copy functor used for locating and marking unique values.
+template <typename Element, bool has_nulls = true>
+struct copy_unique_functor
+{
+    column_device_view d_column;
+    const int32_t* d_ordinals;
+    int32_t* d_indices;
+
+    __device__ bool operator()(int32_t idx)
+    {
+        if( idx==0 )
+        {   // first element is always unique
+            d_indices[0] = 0;
+            return true;
+        }
+        // check if adjacent elements match
+        auto lhs_index = d_ordinals[idx-1];
+        auto rhs_index = d_ordinals[idx];
+        experimental::element_equality_comparator<has_nulls> comparator{d_column,d_column,true};
```

Comment:
Would this benefit from being put in the constructor of this functor?

Reply:
> Would this benefit from being put in the constructor of this functor?

I think that is a good idea. No need to create this in every thread.